File: self pased DSA/Tree/17-Check-Completeness-of-a-Binary-Tree.cpp
/*
Given the root of a binary tree, determine if it is a complete binary tree.

In a complete binary tree, every level, except possibly the last, is completely filled, and all nodes in the last level 
are as far left as possible. It can have between 1 and 2h nodes inclusive at the last level h.
*/


class Solution {
public:
    bool isCompleteTree(TreeNode* root) {
        // use level order traversel
        // bfs

        if(root==NULL) return true;
        queue<TreeNode*> q;
        q.push(root);
        bool f=0;

        while(!q.empty()){
            auto node = q.front();
            q.pop();

            if(node->left){
                if(f==1) return false;
                q.push(node->left);
            }
            else f=1; // mark left node is not present

            if(node->right){
                if(f==1) return false; // previously left node was missing, not a complete Btree
                q.push(node->right);
            }
            else f=1;
        }
        return true;
    }
};

File: code_java/self pased DSA/Tree/17-Check-Completeness-of-a-Binary-Tree.java
/**
 * Problem: Check if a Binary Tree is Complete
 * -------------------------------------------
 * A complete binary tree is defined as a binary tree where every level,
 * except possibly the last, is completely filled, and all nodes in the
 * last level are as far left as possible.
 *
 * Why this matters:
 * Complete binary trees are important in data structures like heaps,
 * where shape properties affect performance.
 *
 * Approach:
 * - Use a level-order traversal (BFS) to visit nodes level-by-level from left to right.
 * - Maintain a flag that indicates if a missing child (null) has been encountered.
 * - Once a missing child is found, all following nodes should not have any children.
 *   If a node after that has children, the tree is not complete.
 *
 * Why this works:
 * In a complete binary tree, nodes must fill every position from left to right
 * without gaps until the last level. Encountering a null child signals the start
 * of possible missing nodes. If any node afterwards has children, it violates completeness.
 *
 * Steps to solve:
 * 1. Initialize a queue and add the root.
 * 2. Traverse nodes level-wise:
 *    - If current node has a left child and flag is false, enqueue left child.
 *      Else, if flag is true but left child exists, return false.
 *    - If current node has a right child and flag is false, enqueue right child.
 *      Else, if flag is true but right child exists, return false.
 *    - If left or right child is missing, set flag true.
 * 3. If traversal finishes without violations, return true.
 *
 * Time Complexity: O(N), N = number of nodes (each node visited once)
 * Space Complexity: O(N) due to queue in worst case
 */

import java.util.LinkedList;
import java.util.Queue;

public class Solution {
    public boolean isCompleteTree(TreeNode root) {
        if (root == null) return true;

        Queue<TreeNode> queue = new LinkedList<>();
        queue.offer(root);
        boolean flag = false; // Indicates a missing child has been seen

        while (!queue.isEmpty()) {
            TreeNode current = queue.poll();

            if (current.left != null) {
                if (flag) {
                    // Previously encountered a missing child, but found a left child now
                    return false;
                }
                queue.offer(current.left);
            } else {
                // Missing left child found, set flag
                flag = true;
            }

            if (current.right != null) {
                if (flag) {
                    // Previously encountered a missing child, but found a right child now
                    return false;
                }
                queue.offer(current.right);
            } else {
                // Missing right child found, set flag
                flag = true;
            }
        }
        return true;
    }
}

// Definition for a binary tree node.
class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;
    TreeNode(int x) { val = x; }
}